### PR TITLE
fix: mongodb the field 'unique' is not valid for _id

### PR DESCRIPTION
**Pull request by fnaranjo-vmw**
[#186981808](https://www.pivotaltracker.com/story/show/186981808)

We started to see some errors related to 'unique' field being invalid for _id key. There is very little clarity online about the issue. More so, if we focus on Azure CosmoDB for MongoDB.

After some trial and error, it seems that mongodb instances created in location 'westus2' have started imposing this, while instances created in location 'westus' work without issues with the same configuration.

Changes in this commit is an attempt to cover our backs and be prepared in case Azure imposes this validation on all locations in the near future.

These changes should be backward compatible without any user intervention. Both, from the point of view of existing plans, and from the point of view of existing instances.
HOWEVER, we may want to introduce an upgrade test to ensure edge cases with default values work as expected.

### Checklist:

* [ ] Have you added Release Notes in the docs repositories?
* [x] Have you followed the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary)?

### CODE REVIEW

File: azure-mongodb.yml [hunk 1]
```diff
@@ -111,6 +111,15 @@ provision:
     type: string
     details: Name for your shard key
     default: uniqueKey
+  - field_name: indexes
+    required: false
+    type: string
+    details: A comma-separated list of non-unique indexes.
```
Comment (blgm) [inline azure-mongodb.yml:117]: This could have been an array rather than a comma-separated list like [AWS DynamoDB Table](https://github.com/cloudfoundry/csb-brokerpak-aws/blob/afe92bf52be957ff388ac8fc514f4e7fb73bef28/aws-dynamodb-table.yml#L67). I don't have a strong feeling about what's best. Maybe DynamoDB was a mistake and a comma-separated list is better. Maybe we should check in with the team on this to make sure we have consensus.

File: terraform-tests/cosmosdb_mongo_test.go [hunk 2]
```diff
@@ -132,7 +134,19 @@ var _ = Describe("CosmosDB Mongo", Label("cosmosdb-mongo-terraform"), Ordered, f
 					"database_name":       Equal(dbName),
 					"default_ttl_seconds": BeNumerically("==", 777),
 					"shard_key":           Equal("key1"),
-				}))
+
+					"index": BeEquivalentTo([]any{
```
Comment (blgm) [inline terraform-tests/cosmosdb_mongo_test.go:138]: Alternatively:
```go
  "index": ConsistOf(
    map[string]any...,
    map[string]any...,
  ),
```

Or:
```go
 "index": ConsistOf(
    MatchAllKeys(
     "keys": ConsistOf("_id"),
     "unique": BeTrue(),
    ),
   ....,
  ),
```

It's debatable which of these is more readable and gives a better message on failure.

File: azure-mongodb.yml (hunk 1 above)
Comment (blgm) [reply azure-mongodb.yml:117]: We discussed and decided that we are happy with the implemented approach.

**Review by blgm — APPROVED** [review]

File: azure-mongodb.yml [hunk 3]
```diff
@@ -111,6 +111,15 @@ provision:
     type: string
     details: Name for your shard key
     default: uniqueKey
+  - field_name: indexes
```
[the remark this answers is not in the document]
Comment (fnaranjo-vmw) [reply azure-mongodb.yml:114]: Good question. When writing the code I checked whether `indexes` was a valid word.
Now I have checked again to compare both spellings. According to Cambridge Dictionary both are valid spellings:
https://dictionary.cambridge.org/dictionary/english/index

> plural indices UK /ˈɪn.dɪ.siːz/ US /ˈɪn.də.siːz/ indexes

File: terraform-tests/cosmosdb_mongo_test.go (hunk 2 above)
Comment (fnaranjo-vmw) [reply terraform-tests/cosmosdb_mongo_test.go:138]: Not being an expert myself in Gomega Matchers,
I tried different combinations of Matchers without success.
`BeEquivalentTo` helped me get the test passing.

I acknowledge, BeEquivalentTo is probably less useful for:
- Large objects (since I guess it will print the whole object if something fails)
- Deep nesting (same as above)
- When more flexibility is needed (it doesn't allow ignoring missing or additional keys)

I believe It has one very big advantage though, when used appropriately:
- It reads exactly as any other golang struct, without requiring any Gomega knowledge.

This is exactly what helped me debug what types I needed exactly.
